### DISCUSSION TITLE
Is automation possible?

### DISCUSSION THREAD

**Discussion by giff-h** [aclap-dev/video-downloadhelper#258] [Help and Q&A]
### Your question

Is it possible to use VDH in JS or something else as a library that can read a page content and return a list of downloadable videos/streams, and start a download when given one of the videos or streams? If it requires a paid license that's fine.

### Addon & CoApp version and details

version: 8.2.1.1
channel: stable
build date: 2024-02-19
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-US
license: unset
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.10",
  "binary": "C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "windows",
    "arch": "x86_64",
    "node": "18"
  },
  "home": "C:\\Users\\hammydotpy",
  "converterBinary": "C:\\Program Files\\DownloadHelper CoApp\\ffmpeg.exe",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "6.0"
}

**Comment by GrampaWildWilly** [comment]
Yeah, it would be kind of cool if there were some APIs exposed to allow some automation, like scheduling a download to start at a given time, stuff like that.  But to the best of my knowledge (I don't work for VDH, I'm just another user) there are no user-accessible APIs to VDH.  What would be extremely useful would be the ability to schedule the recording of a livestreamed TV show.  I'm thinking sporting events, but I suppose any TV broadcast would work the same.  But typically, the URL of the stream is not published by the broadcaster until a few minutes before the opening faceoff, the first tee shot, the first serve, the opening kickoff, etc.  It might be a bit of a problem to deal with such cases.  Still, it might be feasible.

Just as an aside, you are using a woefully out of date CoApp.  You can find the latest CoApp -- and many other useful things -- by clicking the Wiki link at the top of this page. The Wiki page shows a Table of Contents down the right side. You would do well to look at all of the chapters, but for the moment, just click the chapter on CoApp Installation. You should be able to take it from there.

**Reply by giff-h** [reply]
yes I realized that when I went to copy the info, it's up to date now 👍

**Comment by paulrouget** [comment]
Yeah please let me know how you would picture this.

I can explore the idea and see what's doable.

**Reply by GrampaWildWilly** [reply]
I am so glad you are considering this.

OK.  Let's start here:

https://www.nbcsports.com/watch/golf

I believe anybody can visit that page & it will look like this:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/d7750fa9-475a-41b2-aa10-c0cab65ee3e9)

On the left side you can see what's going to be livestreamed next.  In this case it's a ladies' golf tournament in China, to start at 11:00 PM tonight in my time zone.  You're going to have a choice between Peacock & Golf Channel.  Right now at the moment I took that screenshot (time stamp in the bottom right corner) those things on the left are just text strings.  Closer to air time, they will be active links.  Closer to air time seems to be a fluid concept.  It might be 1 hour before air time, a half hour, 15 minutes, 5 minutes, I think I've seen all of those cases.  But at some point a bit before 11:00 PM tonight, there will be clickable links there.  You need to have a cable TV subscription that includes Golf Channel in order to actually get the livestream.  Peacock is a separate subscription service.  I don't know if it's available globally.  I have Golf Channel in my cable TV lineup so I will be able to click the Golf Channel link tonight & watch the tournament.  Now, as it happens, this content is behind DRM-protection so recording it is not on the cards in any case.  But this illustrates the kind of thing I'd like to be able to do.

Here's another one:

https://www.medici.tv/en/concerts/rafael-payare-conducts-beethoven-first-symphony-shostakovich-eighth-symphony-orchestre-symphonique-montreal

At the moment, the page looks like this:

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/11a472dd-2a8f-42cb-bdf2-ec259aea3c8d)

Once again, at some time a little after 7:00PM this coming Wednesday, that image will transform into a video player.  Although this content does require a login to view, Medici has both free & paid subscriptions.  With a free subscription, it is usually possible to record their livestreams as long as you tune in while the livestream is actually live.  And Medici content is not protected by DRM.  VDH can record this stuff.  They are pretty box standard, too, in the way they use HLS.  They have very ordinary master manifests that are totally amenable to access by VDH.  But once again, we have the issue of there being no URL on that page until close to young Rafael there taking the podium at La Place des Arts this coming Wednesday evening.

There is a big difference between NBC Sports & Medici TV.  At the appointed time on the NBC page I've shown above, a link will appear.  Clicking that link will open a new page on which the video player will reside.  So there's a level of indirection.  But the Medici page I've shown above IS the page on which you will be able to play the livestream.  So it's more direct.  For NBC, we don't know the URL of the web page where the livestream will be hosted & won't know until just before air time.  For Medici, we already know right now where the livestream will be hosted.

So I would want to tell VDH something like this:
- Here's the URL of a page on which a livestream will be hosted.
- At a specified future time, open that page, find the livestream, & record it to a directory & file name I specify.  There might also need to be options for specifying the desired resolution to record.

That actually sounds pretty straightforward for Medici TV.  For NBC Sports, it might not be possible because of the indirection.

Let's hear from others, too.

**Reply by giff-h** [reply]
Here's a mock-up of what I'd think could be available, in Typescript (The interfaces are not based on anything concrete, just available for some context):

```Typescript
// vdh.ts

interface DownloadableVideo {
  // w,h tuple
  resolution: [number, number];

  // bits/sec
  bitrate: number;

  // Identifying URI
  mediaLink: string;
}

/**
 * Given a document node or URL, scans the page for streams and returns a list of available streams to download.
 *
 * If the given page is a node, scans it as is. If it's a URL, downloads it immediately and scans the resulting page.
 */
export function getStreamsFromPage(documentOrUrl: Node | string): DownloadableVideo[] {
  // implementation
}

interface RunningDownload {
  // Returns the current status of the download
  getStatus: () => {
    // Seconds (milliseconds?) downloaded so far. Video runtime, not time spent downloading.
    videoLength: number;

    // Bytes downloaded so far
    videoSize: number;
  }

  // Stops the download immediately, leaving a playable video file
  cancel: () => void;
}

/**
 * Begins downloading the given video stream, saving to the given target file, and calls the given callback when done.
 *
 * If the optional failed callback is given, and the download fails, that callback is called, and the done callback is not called.
 * If the optional failed callback is NOT given, the done callback is called on failure.
 */
export function downloadStream(stream: DownloadableVideo, targetFilepath: string, done: () => void, failed?: (reason: string) => void): RunningDownload {
  // implementation
}
```

And how it could be used, to read a webpage every half hour and if there's a stream available, start downloading the highest resolution one:

```Typescript
// myscript.ts
import { interval } from "rxjs";
import { downloadStream, getStreamsFromPage } from "vdh";

const thirtyMinutesInMS = 1000 * 60 * 30;
let runningDownload;

interval(thirtyMinutesInMS).subscribe(() => {
  if (!runningDownload) {
    const streams = getStreamsFromPage(URL);
    if (streams.length) {
      streams.sort((a, b) => b.resolution[1] - a.resolution[1]);
      runningDownload = downloadStream(streams[0], "/path/to/file", () => { runningDownload = null; });
    }
  }
});
```

**Reply by paulrouget** [reply]
That's useful.

Adding that to my todolist. Could be real cool, and not that hard to put together.

Probably not gonna work on that immediately, lot of other "normal user" stuff to fix first, but that sounds like something that could be real useful.

**Reply by GrampaWildWilly** [reply]
I'm assuming that if the API is available to that programming language that giff is using that I will be able to issue plain old CALL commands from a Windows command prompt & get similar effects.

**Reply by paulrouget** [reply]
We have discussed in the past of having some sort of greasemonkey scripts for VDH. Basically, a TS or JS code editor in VDH that would allow loading scripts within VDH, to add hooks or trigger scheduled actions.

**Reply by issgazer** [reply]
I'm looking to record a scheduled live YouTube broadcast. It's weekly at the same time.  The url for the show is available 6 days prior to the show. Went looking for a solution and found this conversation. Just want to show my interest in the possibilities your discussing. Following!
